llvm/lib/Target/WebAssembly/WebAssemblyISelLowering.cpp

```diff
@@ -2143,6 +2143,32 @@ WebAssemblyTargetLowering::LowerSIGN_EXTEND_INREG(SDValue Op,
                      Op.getOperand(1));
 }
 
+static SDValue GetExtendHigh(SDValue Op, unsigned UserOpc, EVT VT,
+                             SelectionDAG &DAG) {
+  if (Op.getOpcode() != ISD::VECTOR_SHUFFLE)
+    return SDValue();
+
+  assert((UserOpc == WebAssemblyISD::EXTEND_LOW_U ||
+          UserOpc == WebAssemblyISD::EXTEND_LOW_S) &&
+         "expected extend_low");
+  auto *Shuffle = cast<ShuffleVectorSDNode>(Op.getNode());
+  ArrayRef<int> Mask = Shuffle->getMask();
+
+  // Look for a shuffle which moves from the high half to the low half.
+  size_t FirstIdx = Mask.size() / 2;
+  for (size_t i = 0; i < Mask.size() / 2; ++i) {
```

Comment:
Do we have to check that the top half of the mask has expected values as well?

Reply:
We don't, but as the user is an `extend_low`, the high half of the result doesn't matter. But I will add a check that the extend is the only user.

Reply:
I implemented and then removed the user check as I couldn't see the benefit.